## Import Libraries

In [1]:
### Direct Nearest Timestamp Matching ###
import pandas as pd
import numpy as np
import struct
import os
import matplotlib.pyplot as plt
import datetime

from data_utils import get_traces
from beam_settings_parser_hdf5 import BeamConfigParserHDF5
from beam_settings_prep import BeamConfigPreProcessor


import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

ModuleNotFoundError: No module named 'data_utils'

## Load and Merge Beam Position Monitor (BPM) datasets

In [3]:
### FebMar22 .csv beam settings
beam_settings_data_path = "/work/data_science/suf_sns/beam_configurations_data/processed_data/clean_beam_config_processed_df.csv"
configs = pd.read_csv(beam_settings_data_path)
configs = configs.drop("Unnamed: 0", axis=1, errors='ignore')
configs['timestamps'] = pd.to_datetime(configs['timestamps'])

### Sep24 hdf5 beam settings
beam_param_parser_cfg = {"data_location": "/work/data_science/suf_sns/beam_configurations_data/hdf5_sept2024/"}
parser = BeamConfigParserHDF5(beam_param_parser_cfg)
data, _ = parser.run()

beam_settings_prep_cfg = {
    "rescale": False,
    "beam_config": ['FE_IS:Match:TunerPos',
                    'LEBT:Chop_N:V_Set',
                    'LEBT:Chop_P:V_Set',
                    'LEBT:Focus_1:V_Set',
                    'LEBT:Focus_2:V_Set',
                    'LEBT:Steer_A:V_Set',
                    'LEBT:Steer_B:V_Set',
                    'LEBT:Steer_C:V_Set',
                    'LEBT:Steer_D:V_Set',
                    'Src:Accel:V_Set',
                    'Src:H2:Flw_Set',
                    'Src:Ign:Pwr_Set',
                    'Src:RF_Gnd:Pwr_Set',
                    'ICS_Chop:RampDown:PW', # ICS_Chop-RampDown-PW
                    'ICS_Chop:RampUp:PWChange', # ICS_Chop-RampUp-PWChange
                    'ICS_MPS:Gate_Source:Offset', # ICS_MPS-Gate_Source-Offset
                    'ICS_Tim:Chop_Flavor1:BeamOn', # ICS_Chop-BeamOn-Width
                    'ICS_Tim:Chop_Flavor1:OnPulseWidth', # ICS_Chop-BeamOn-PW
                    'ICS_Tim:Chop_Flavor1:RampUp', # ICS_Chop-RampUp-Width
                    'ICS_Tim:Chop_Flavor1:StartPulseWidth', # ICS_Chop-RampUp-PW
                    'ICS_Tim:Gate_BeamRef:GateWidth', # ICS_Tim-Gate_BeamRef-GateWidth
                    'ICS_Tim:Gate_BeamOn:RR']
}
prep = BeamConfigPreProcessor(beam_settings_prep_cfg)
prepared_settings, run_cfg = prep.run(data)

Provided file is not hdf5 format, skipping:  .ipynb_checkpoints
BeamParamParser: Number of samples parsed
FE_IS:Match:TunerPos 633
ICS_Chop:RampDown:PW 41
ICS_Chop:RampUp:PWChange 869
ICS_MPS:Gate_Source:Offset 262
ICS_Tim:Chop_Flavor1:BeamOn 1780
ICS_Tim:Chop_Flavor1:OnPulseWidth 2564
ICS_Tim:Chop_Flavor1:RampUp 52
ICS_Tim:Chop_Flavor1:StartPulseWidth 16
ICS_Tim:Gate_BeamOn:RR 12437
ICS_Tim:Gate_BeamRef:GateWidth 1818
LEBT:Chop_N:V_Set 6557
LEBT:Chop_P:V_Set 6561
LEBT:Focus_1:V_Set 3018
LEBT:Focus_2:V_Set 3018
LEBT:Steer_A:V_Set 6550
LEBT:Steer_B:V_Set 6557
LEBT:Steer_C:V_Set 6554
LEBT:Steer_D:V_Set 6551
Src:Accel:V_Set 3408
Src:H2:Flw_Set 2959
Src:Ign:Pwr_Set 2942
Src:RF_Gnd:Pwr_Set 5
['FE_IS:Match:TunerPos', 'LEBT:Chop_N:V_Set', 'LEBT:Chop_P:V_Set', 'LEBT:Focus_1:V_Set', 'LEBT:Focus_2:V_Set', 'LEBT:Steer_A:V_Set', 'LEBT:Steer_B:V_Set', 'LEBT:Steer_C:V_Set', 'LEBT:Steer_D:V_Set', 'Src:Accel:V_Set', 'Src:H2:Flw_Set', 'Src:Ign:Pwr_Set', 'Src:RF_Gnd:Pwr_Set', 'ICS_Chop:RampDown:PW', 'IC

/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  master_df.fillna(method='ffill', inplace=True)
/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  master_df.fillna(method='ffill', inplace=True)
/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  master_df.fillna(method='ffill', inplace=True)
/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  master_df.fillna(method='ffill', inplace=True)
/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and w

Length of beam param df:  48909


/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  master_df.fillna(method='ffill', inplace=True)
/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  master_df.fillna(method='ffill', inplace=True)
/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  master_df.fillna(method='ffill', inplace=True)
/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  master_df.fillna(method='ffill', inplace=True)
/home/brodaj/beam_settings_prep.py:87: FutureWarning: DataFrame.fillna with 'method' is deprecated and w

In [4]:
def configs_hist(dataframe, timestamp):
    subset_columns = dataframe.columns.tolist()
    subset_columns.remove(timestamp)
    df_shifted = dataframe[subset_columns].shift(1)
    mask = (dataframe[subset_columns] == df_shifted).all(axis=1)
    dataframe = dataframe[~mask] #.reset_index(drop=True)
    
    dataframe['time_diff'] = dataframe[timestamp].diff()
    dataframe['timestamps_trm'] = dataframe[timestamp] + dataframe[timestamp].diff().shift(-1) - datetime.timedelta(seconds=0.000001)
    
    subset_columns.insert(0, timestamp)
    subset_columns.insert(1, "timestamps_trm")
    
    return dataframe[subset_columns]

def summary(text, df):
    print(f'{text} shape: {df.shape}')
    summ = pd.DataFrame(df.dtypes, columns=['dtypes'])
    summ['null'] = df.isnull().sum()
    summ['unique'] = df.nunique()
    summ['min'] = df.min()
    summ['median'] = df.median()
    summ['max'] = df.max()
    summ['mean'] = df.mean()
    summ['std'] = df.std()
    summ['duplicate'] = df.duplicated().sum()
    return summ

In [5]:
# configs = configs_hist(configs,"timestamps")
# prepared_settings = configs_hist(prepared_settings,"timestamps")

In [6]:
column_to_add = [
    'FE_IS:Match:TunerPos',
    'LEBT:Chop_N:V_Set',
    'LEBT:Chop_P:V_Set',
    'LEBT:Focus_1:V_Set',
    'LEBT:Focus_2:V_Set',
    'LEBT:Steer_A:V_Set',
    'LEBT:Steer_B:V_Set',
    'LEBT:Steer_C:V_Set',
    'LEBT:Steer_D:V_Set',
    'Src:Accel:V_Set',
    'Src:H2:Flw_Set',
    'Src:Ign:Pwr_Set',
    'Src:RF_Gnd:Pwr_Set',
    'ICS_Tim:Gate_BeamOn:RR'
]

for entry in column_to_add:
    configs[entry] = np.nan

configs = configs.rename(columns={
    'ICS_Chop-RampDown-PW' : 'ICS_Chop:RampDown:PW', 
    'ICS_Chop-RampUp-PWChange' : 'ICS_Chop:RampUp:PWChange',
    'ICS_MPS-Gate_Source-Offset' : 'ICS_MPS:Gate_Source:Offset',
    'ICS_Chop-BeamOn-Width' : 'ICS_Tim:Chop_Flavor1:BeamOn',
    'ICS_Chop-BeamOn-PW' : 'ICS_Tim:Chop_Flavor1:OnPulseWidth',
    'ICS_Chop-RampUp-Width' : 'ICS_Tim:Chop_Flavor1:RampUp',
    'ICS_Chop-RampUp-PW' : 'ICS_Tim:Chop_Flavor1:StartPulseWidth',
    'ICS_Tim-Gate_BeamRef-GateWidth' : 'ICS_Tim:Gate_BeamRef:GateWidth', 
    })

In [7]:
beam_config = [
    'timestamps',
    'FE_IS:Match:TunerPos',
    'LEBT:Chop_N:V_Set',
    'LEBT:Chop_P:V_Set',
    'LEBT:Focus_1:V_Set',
    'LEBT:Focus_2:V_Set',
    'LEBT:Steer_A:V_Set',
    'LEBT:Steer_B:V_Set',
    'LEBT:Steer_C:V_Set',
    'LEBT:Steer_D:V_Set',
    'Src:Accel:V_Set',
    'Src:H2:Flw_Set',
    'Src:Ign:Pwr_Set',
    'Src:RF_Gnd:Pwr_Set',
    'ICS_Chop:RampDown:PW', # ICS_Chop-RampDown-PW
    'ICS_Chop:RampUp:PWChange', # ICS_Chop-RampUp-PWChange
    'ICS_MPS:Gate_Source:Offset', # ICS_MPS-Gate_Source-Offset
    'ICS_Tim:Chop_Flavor1:BeamOn', # ICS_Chop-BeamOn-Width
    'ICS_Tim:Chop_Flavor1:OnPulseWidth', # ICS_Chop-BeamOn-PW
    'ICS_Tim:Chop_Flavor1:RampUp', # ICS_Chop-RampUp-Width
    'ICS_Tim:Chop_Flavor1:StartPulseWidth', # ICS_Chop-RampUp-PW
    'ICS_Tim:Gate_BeamRef:GateWidth', # ICS_Tim-Gate_BeamRef-GateWidth
    'ICS_Tim:Gate_BeamOn:RR']

In [8]:
configs = configs.loc[:, beam_config]

In [9]:
bpm = pd.concat([configs, prepared_settings])

/scratch/slurm/41600596/.cache/tmp/ipykernel_1320317/1496517970.py:1: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  bpm = pd.concat([configs, prepared_settings])


In [10]:
summary('bpm_summary', bpm)

bpm_summary shape: (67764, 23)


dtypes   null  unique  \
timestamps                            datetime64[ns]      0   66937   
FE_IS:Match:TunerPos                         float64  18855       3   
LEBT:Chop_N:V_Set                            float64  18855       2   
LEBT:Chop_P:V_Set                            float64  18855       4   
LEBT:Focus_1:V_Set                           float64  18855       5   
LEBT:Focus_2:V_Set                           float64  18855       4   
LEBT:Steer_A:V_Set                           float64  18855       1   
LEBT:Steer_B:V_Set                           float64  18855       1   
LEBT:Steer_C:V_Set                           float64  18855       1   
LEBT:Steer_D:V_Set                           float64  18855       1   
Src:Accel:V_Set                              float64  18855       7   
Src:H2:Flw_Set                               float64  18855      12   
Src:Ign:Pwr_Set                              float64  18855       4   
Src:RF_Gnd:Pwr_Set                           float64  18855       1   
ICS_Chop:RampDown:PW                         float64      0      16   
ICS_Chop:RampUp:PWChange                     float64      0      27   
ICS_MPS:Gate_Source:Offset                   float64      0     140   
ICS_Tim:Chop_Flavor1:BeamOn                  float64      0     645   
ICS_Tim:Chop_Flavor1:OnPulseWidth            float64      0      39   
ICS_Tim:Chop_Flavor1:RampUp                  float64      0      49   
ICS_Tim:Chop_Flavor1:StartPulseWidth         float64      0      36   
ICS_Tim:Gate_BeamRef:GateWidth               float64      0     664   
ICS_Tim:Gate_BeamOn:RR                       float32  18855      65   

                                                             min  \
timestamps                            2021-02-11 08:18:29.948852   
FE_IS:Match:TunerPos                                   58.009998   
LEBT:Chop_N:V_Set                                           2.43   
LEBT:Chop_P:V_Set                                           0.54   
LEBT:Focus_1:V_Set                                        2.8848   
LEBT:Focus_2:V_Set                                           6.0   
LEBT:Steer_A:V_Set                                           1.7   
LEBT:Steer_B:V_Set                                           2.1   
LEBT:Steer_C:V_Set                                           1.6   
LEBT:Steer_D:V_Set                                           2.0   
Src:Accel:V_Set                                             18.0   
Src:H2:Flw_Set                                              3.22   
Src:Ign:Pwr_Set                                            104.0   
Src:RF_Gnd:Pwr_Set                                         4.717   
ICS_Chop:RampDown:PW                                         0.0   
ICS_Chop:RampUp:PWChange                                     0.0   
ICS_MPS:Gate_Source:Offset                            -949.96748   
ICS_Tim:Chop_Flavor1:BeamOn                                  0.0   
ICS_Tim:Chop_Flavor1:OnPulseWidth                            0.0   
ICS_Tim:Chop_Flavor1:RampUp                                  0.0   
ICS_Tim:Chop_Flavor1:StartPulseWidth                         0.0   
ICS_Tim:Gate_BeamRef:GateWidth                               0.0   
ICS_Tim:Gate_BeamOn:RR                                       0.1   

                                                             median  \
timestamps                            2024-07-16 09:15:27.486576640   
FE_IS:Match:TunerPos                                      58.040001   
LEBT:Chop_N:V_Set                                               2.7   
LEBT:Chop_P:V_Set                                               2.7   
LEBT:Focus_1:V_Set                                             46.0   
LEBT:Focus_2:V_Set                                             39.0   
LEBT:Steer_A:V_Set                                              1.7   
LEBT:Steer_B:V_Set                                              2.1   
LEBT:Steer_C:V_Set                                              1.6  

## Load and Merge Beam Tracings from Differential Current Monitor (DCM)

### Feb/Mar22 Data

In [11]:
dataset1_loc = "/work/data_science/suf_sns/DCM_Errant/"

start_date = 20220218
end_date = 20220318

anomaly_type = "00110000" #--  48
length_of_waveform = 10000

exclude_dates = [20220220, 20220221, 20220222, 20220223, 20220301, 20220308, 20220309, 2022015]

In [12]:
filtered_normal_files = []
filtered_anomaly_files = []
for root, subfolders, files in os.walk(dataset1_loc):
    for file in files:
        if '.bin' in file and 'DCML' in file:
            try:
                date = int(file[:8])
            except:
                print("Error in filename: ", file)
                print("-- Date could not be read!")
                continue
            if date >= start_date and date <= end_date:

                if date in exclude_dates:
                    continue

                if '00000000' in file:
                    filtered_normal_files.append(os.path.join(root, file))
                elif anomaly_type in file:
                    filtered_anomaly_files.append(os.path.join(root, file))

print('Number of available normal files: ', len(filtered_normal_files))
print('Number of available anomaly files: ', len(filtered_anomaly_files))

Number of available normal files:  2173
Number of available anomaly files:  2953


Number of available normal files: 2173\
Number of available anomaly files: 2953\
Total files to work with: **5126**

**Question** - Why is the length of the traces 26, while the length of timestamps 27?\
*Thought* --> the first array in the traces has "unusual data" compared to the rest, we were supplied code that ignored the first record of the traces, should we do the same for the timestamps?

### Sept24 Data

In [13]:
dataset2_loc = "/w/data_science-sciwork24/suf_sns/DCML_dataset_Sept2024"

In [14]:
filtered_normal_files2 = []
filtered_anomaly_files2 = []
subfolders = [ f.path for f in os.scandir(dataset2_loc) if f.is_dir() ]
for directory in subfolders:
    if "normal" in directory or "anomal" in directory:                
        for root, subfolders, files in os.walk(directory):
            for file in files:
                full_path = root
                if ".gz" in file:
                    if 'normal' in directory:
                        filtered_normal_files2.append(os.path.join(full_path, file))
                    elif "anomal" in directory:
                        filtered_anomaly_files2.append(os.path.join(full_path, file))
                        
print('Number of available normal files: ', len(filtered_normal_files2))
print('Number of available anomaly files: ', len(filtered_anomaly_files2))

Number of available normal files:  10699
Number of available anomaly files:  20592


In [15]:
index = np.random.randint(0, len(filtered_normal_files))
filepath = filtered_normal_files[index]
print(index)
try:
    traces, timestamps = get_traces(filepath, var_id="Trace2", begin=3000, shift=length_of_waveform, data_type=0)
except Exception as e:
    traces = []
    timestamps = []
    print("Error in reading the file: ", filepath)
    print("Error:", e)

1041


In [16]:
# import tensorflow as tf

### Merge Two datasets together

Can we switch from list to dictionary?

In [28]:
traces = []
timestamps = []
flag = []
file = []
for dcml in filtered_normal_files[:10]:
    tmp_trace, tmp_timestamp = get_traces(dcml, var_id="Trace2", begin=3000, shift=length_of_waveform, data_type=0)
    tmp_trace = np.array(tmp_trace[1:])
    tmp_timestamp = np.array(tmp_timestamp[1:])
    for sample in tmp_trace:
        # sample = tf.convert_to_tensor(sample)
        traces.append(sample)
        flag.append(0)
        file.append('FebMar22')
    for time in tmp_timestamp:
        timestamps.append(time)

for dcml in filtered_anomaly_files[:10]:
    tmp_trace, tmp_timestamp = get_traces(dcml, var_id="Trace2", begin=3000, shift=length_of_waveform, data_type=-1)
    for sample in tmp_trace:
        # sample = tf.convert_to_tensor(sample)
        traces.append(sample)
        flag.append(1)
        file.append('FebMar22')
    for time in tmp_timestamp:
        timestamps.append(time)

for dcml in filtered_normal_files2[:10]:
    tmp_trace, tmp_timestamp = get_traces(dcml, var_id="Trace2", begin=3000, shift=length_of_waveform, data_type=0)
    tmp_trace = np.array(tmp_trace[1:])
    tmp_timestamp = np.array(tmp_timestamp[1:])
    for sample in tmp_trace:
        traces.append(sample)
        flag.append(0)
        file.append('Sep24')
    for time in tmp_timestamp:
        timestamps.append(time)

for dcml in filtered_anomaly_files2[:10]:
    tmp_trace, tmp_timestamp = get_traces(dcml, var_id="Trace2", begin=3000, shift=length_of_waveform, data_type=-1, alarm=48)
    for sample in tmp_trace:
        traces.append(sample)
        flag.append(1)
        file.append('Sep24')
    for time in tmp_timestamp:
        timestamps.append(time)

File:  /w/data_science-sciwork24/suf_sns/DCML_dataset_Sept2024/anomalies/03_18/20240903_183444.5928_DCML.bin.gz   bWidth below threshold: 52
File:  /w/data_science-sciwork24/suf_sns/DCML_dataset_Sept2024/anomalies/03_18/20240903_182423.6166_DCML.bin.gz   bWidth below threshold: 30


There should be **698,438** in each array

&nbsp; &nbsp; &nbsp;2173 * 26 = 56,498\
10,699 * 60 = 641,940

Actauls equals to **xxxxx**, due to drops from Non-zero alarms present in the data

In [29]:
dcm = pd.DataFrame({'anomoly_flag':flag, 'file':file, 'timestamps':timestamps, 'traces':traces})

## Join DCM & BPM into single DataFrame

In [30]:
merged_df = pd.merge_asof(
    dcm.sort_values("timestamps"), 
    bpm.sort_values("timestamps"), 
    on="timestamps", 
    direction="nearest"
)

In [31]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 864 entries, 0 to 863
Data columns (total 26 columns):
 #   Column                                Non-Null Count  Dtype         
---  ------                                --------------  -----         
 0   anomoly_flag                          864 non-null    int64         
 1   file                                  864 non-null    object        
 2   timestamps                            864 non-null    datetime64[ns]
 3   traces                                864 non-null    object        
 4   FE_IS:Match:TunerPos                  594 non-null    float64       
 5   LEBT:Chop_N:V_Set                     594 non-null    float64       
 6   LEBT:Chop_P:V_Set                     594 non-null    float64       
 7   LEBT:Focus_1:V_Set                    594 non-null    float64       
 8   LEBT:Focus_2:V_Set                    594 non-null    float64       
 9   LEBT:Steer_A:V_Set                    594 non-null    float64       
 10  LE

In [32]:
merged_df

anomoly_flag      file                 timestamps  \
0               0  FebMar22 2022-02-28 08:30:00.586558   
1               0  FebMar22 2022-02-28 08:30:00.603224   
2               0  FebMar22 2022-02-28 08:30:00.619889   
3               0  FebMar22 2022-02-28 08:30:00.636554   
4               0  FebMar22 2022-02-28 08:30:00.653220   
..            ...       ...                        ...   
859             0     Sep24 2024-09-18 10:57:38.721789   
860             0     Sep24 2024-09-18 10:57:38.738455   
861             0     Sep24 2024-09-18 10:57:38.755121   
862             0     Sep24 2024-09-18 10:57:38.771787   
863             0     Sep24 2024-09-18 10:57:38.788453   

                                                traces  FE_IS:Match:TunerPos  \
0    [0.00063476566, 0.0009765625, -9.765625e-05, -...                   NaN   
1    [0.000390625, 9.765625e-05, -0.000390625, -0.0...                   NaN   
2    [-9.765625e-05, 0.00024414062, -0.00048828125,...                   NaN   
3    [-0.00029296876, 0.00068359374, 0.00034179687,...                   NaN   
4    [0.00014648438, -4.8828126e-05, -0.00078125, -...                   NaN   
..                                                 ...                   ...   
859  [-9.765625e-05, -4.8828126e-05, 0.0005371094, ...             58.029999   
860  [0.0005371094, 0.0, 4.8828126e-05, -0.00068359...             58.029999   
861  [0.0001953125, -0.00083007815, 0.00092773436, ...             58.029999   
862  [-0.00014648438, 0.00029296876, 0.000390625, -...             58.029999   
863  [-0.0001953125, 0.0, 0.00034179687, -0.0011718...             58.029999   

     LEBT:Chop_N:V_Set  LEBT:Chop_P:V_Set  LEBT:Focus_1:V_Set  \
0                  NaN                NaN                 NaN   
1                  NaN                NaN                 NaN   
2                  NaN                NaN                 NaN   
3                  NaN                NaN                 NaN   
4                  NaN                NaN                 NaN   
..                 ...                ...                 ...   
859                2.7                2.7                46.0   
860                2.7                2.7                46.0   
861                2.7                2.7                46.0   
862                2.7                2.7                46.0   
863                2.7                2.7                46.0   

     LEBT:Focus_2:V_Set  LEBT:Steer_A:V_Set  ...  Src:RF_Gnd:Pwr_Set  \
0                   NaN                 NaN  ...                 NaN   
1                   NaN                 NaN  ...                 NaN   
2                   NaN                 NaN  ...                 NaN   
3                   NaN                 NaN  ...                 NaN   
4                   NaN                 NaN  ...                 NaN   
..                  ...                 ...  ...                 ...   
859                39.0                 1.7  ...               4.717   
860                39.0                 1.7  ...               4.717   
861                39.0                 1.7  ...               4.717   
862                39.0                 1.7  ...               4.717   
863                39.0                 1.7  ...               4.717   

     ICS_Chop:RampDown:PW  ICS_Chop:RampUp:PWChange  \
0                    18.0                       2.0   
1                    18.0                       2.0   
2                    18.0                       2.0   
3                    18.0                       2.0   
4                    18.0                       2.0   
..                    ...                       ...   
859                  18.0                       6.0   
860                  18.0                       6.0   
861                  18.0                       6.0   
862                  18.0                       6.0   
863                  18.0                       6.0   

     ICS_MPS:Gate_Source:Offset  ICS_Tim

In [33]:
merged_df.groupby(['anomoly_flag','file'])['traces'].count()

anomoly_flag  file    
0             FebMar22    260
              Sep24       590
1             FebMar22     10
              Sep24         4
Name: traces, dtype: int64

In [37]:
summary('test',bpm)

test shape: (67764, 23)


dtypes   null  unique  \
timestamps                            datetime64[ns]      0   66937   
FE_IS:Match:TunerPos                         float64  18855       3   
LEBT:Chop_N:V_Set                            float64  18855       2   
LEBT:Chop_P:V_Set                            float64  18855       4   
LEBT:Focus_1:V_Set                           float64  18855       5   
LEBT:Focus_2:V_Set                           float64  18855       4   
LEBT:Steer_A:V_Set                           float64  18855       1   
LEBT:Steer_B:V_Set                           float64  18855       1   
LEBT:Steer_C:V_Set                           float64  18855       1   
LEBT:Steer_D:V_Set                           float64  18855       1   
Src:Accel:V_Set                              float64  18855       7   
Src:H2:Flw_Set                               float64  18855      12   
Src:Ign:Pwr_Set                              float64  18855       4   
Src:RF_Gnd:Pwr_Set                           float64  18855       1   
ICS_Chop:RampDown:PW                         float64      0      16   
ICS_Chop:RampUp:PWChange                     float64      0      27   
ICS_MPS:Gate_Source:Offset                   float64      0     140   
ICS_Tim:Chop_Flavor1:BeamOn                  float64      0     645   
ICS_Tim:Chop_Flavor1:OnPulseWidth            float64      0      39   
ICS_Tim:Chop_Flavor1:RampUp                  float64      0      49   
ICS_Tim:Chop_Flavor1:StartPulseWidth         float64      0      36   
ICS_Tim:Gate_BeamRef:GateWidth               float64      0     664   
ICS_Tim:Gate_BeamOn:RR                       float32  18855      65   

                                                             min  \
timestamps                            2021-02-11 08:18:29.948852   
FE_IS:Match:TunerPos                                   58.009998   
LEBT:Chop_N:V_Set                                           2.43   
LEBT:Chop_P:V_Set                                           0.54   
LEBT:Focus_1:V_Set                                        2.8848   
LEBT:Focus_2:V_Set                                           6.0   
LEBT:Steer_A:V_Set                                           1.7   
LEBT:Steer_B:V_Set                                           2.1   
LEBT:Steer_C:V_Set                                           1.6   
LEBT:Steer_D:V_Set                                           2.0   
Src:Accel:V_Set                                             18.0   
Src:H2:Flw_Set                                              3.22   
Src:Ign:Pwr_Set                                            104.0   
Src:RF_Gnd:Pwr_Set                                         4.717   
ICS_Chop:RampDown:PW                                         0.0   
ICS_Chop:RampUp:PWChange                                     0.0   
ICS_MPS:Gate_Source:Offset                            -949.96748   
ICS_Tim:Chop_Flavor1:BeamOn                                  0.0   
ICS_Tim:Chop_Flavor1:OnPulseWidth                            0.0   
ICS_Tim:Chop_Flavor1:RampUp                                  0.0   
ICS_Tim:Chop_Flavor1:StartPulseWidth                         0.0   
ICS_Tim:Gate_BeamRef:GateWidth                               0.0   
ICS_Tim:Gate_BeamOn:RR                                       0.1   

                                                             median  \
timestamps                            2024-07-16 09:15:27.486576640   
FE_IS:Match:TunerPos                                      58.040001   
LEBT:Chop_N:V_Set                                               2.7   
LEBT:Chop_P:V_Set                                               2.7   
LEBT:Focus_1:V_Set                                             46.0   
LEBT:Focus_2:V_Set                                             39.0   
LEBT:Steer_A:V_Set                                              1.7   
LEBT:Steer_B:V_Set                                              2.1   
LEBT:Steer_C:V_Set                                              1.6  